### PR TITLE
[pkg] reproducible windows installer for bitmask_client

### PR DESCRIPTION
**Pull request by PaixuAabuizia**
provide a environment that allows automated builds of windows installers

- prepare dockerized environment with wine, python, openssl, zlib and mingw
  to build windows binaries from python sourcecode
- prepare dockerized environment with nullsoft installer
  to build installers from binaries
- configure pyinstaller to build binaries
- configure nsis to build distributable executables for bitmask
- configure make all in pkg/windows that results in installers
- add documentation
- ico conversion from data/images
- avoid polluting / in docker image
- install dirspec and copy to wine env
- remove obsolete comments
- fix python path
- figure out that pip install leap.a and pyinstalling a leap.b does not work - so the build script fixes that
- rename dependencies to pyinstaller and move nsis code to installer
- build openvpn, export the binaries for further processing
- correct openvpn dependencies, fetch tap installer compatible with openvpn just built
- install tap-driver with nsis
- pyinstaller-build: fix mixed mkdir / show errors if there are some
- installer-build: prepare rw-copy, do not expose nsh files
- add openvpn_leap.exe to install directory so it gets picked up by nsis
- use setup.py to install bitmask to site-packages to have a version
- separate build directories for granular make
- copy all openvpn dlls to installer
- die to signal failure to parent makefile
- cache installDependencies for quick turn-arround times
- share openssl version between openvpn and pysqlcipher/other pip builds
- collect files during prepare for installer
- default to eip:false, mail:true
- configuration in pyinstaller-build.sh
- win64 tap drivers need special care getting removed from 32bit nsis
- correct registry key that identifies if we installed TAP
- extract version from git-tree, expose to wine python
- create nsh with version for build installer
- allow clean/dirty version with patches
- cleanup / indent / remove comments

### CODE REVIEW

**Comment by kalikaneko** [comment]
great! you have a sample of the generated installer?

File: pkg/pyinst/bitmask.spec [hunk 1]
```diff
@@ -20,19 +35,12 @@ exe = EXE(pyz,
...
+          console=False,
+          icon='../../data/images/mask-icon.ico')
 coll = COLLECT(exe,
-               a.binaries,
-               a.zipfiles,
-               a.datas,
-               strip=False,
-               upx=True,
-               name='bitmask')
-if sys.platform.startswith("darwin"):
```
Comment (kalikaneko) [inline pkg/pyinst/bitmask.spec]: this shouldn't be removed, if we're going to use the same file for both win and osx builds

**Comment by meskio** [comment]
Running it I hit the following compiler error building sqlite:
```
writing build\temp.win32-2.7\Release\src\_sqlite.def
C:\mingw\bin\gcc.exe -shared -s build\temp.win32-2.7\Release\src\module.o build\temp.win32-2.7\Release\src\connection.o build\temp.win32-2.7\Release\src\cursor.o build\temp.win32-2.7\Release\src\cache.o build\temp.win32-2.7\Release\src\mic
roprotocols.o build\temp.win32-2.7\Release\src\prepare_protocol.o build\temp.win32-2.7\Release\src\statement.o build\temp.win32-2.7\Release\src\util.o build\temp.win32-2.7\Release\src\row.o build\temp.win32-2.7\Release\amalgamation\sqlite3
.o build\temp.win32-2.7\Release\src\_sqlite.def -Lc:\openssl\lib -Lc:\openssin -Lc:\zlib\lib -Lc:\zliin -LC:\Python27\libs -LC:\Python27\PCbuild -LC:\Python27\PC\VS9.0 -lpython27 -lmsvcr90 -o build\lib.win32-2.7\pysqlcipher\_sqlite.pyd lib
eay32.lib /LIBPATH:C:\lib
gcc: error: libeay32.lib: No such file or directory
gcc: error: /LIBPATH:C:\lib: No such file or directory
error: command 'C:\\mingw\\bin\\gcc.exe' failed with exit status 1
die: setup.py for pysqlcipher failed
Makefile:12: recipe for target 'pyinstaller' failed
make[1]: *** [pyinstaller] Error 1
```

**Comment by PaixuAabuizia** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

Hi Ruben,

pysqlcipher is not prepared to be build by mingw, the pull request
contains a rough patch: (pyinstaller.sh:174) that was not fenced with
a || die 'patch pysqlcipher setup.py failed'

before the pull request i cleaned up the patch, removing comments, and
missed to recalculate the -@192,10 @192,7 line in the patch head.

the pull request is corrected

Cheers
Paixu
On 02/16/2016 07:22 PM, Ruben Pollan wrote:
> Running it I hit the following compiler error building sqlite:
> 
> |writing build\temp.win32-2.7\Release\src\_sqlite.def 
> C:\mingw\bin\gcc.exe -shared -s 
> build\temp.win32-2.7\Release\src\module.o 
> build\temp.win32-2.7\Release\src\connection.o 
> build\temp.win32-2.7\Release\src\cursor.o 
> build\temp.win32-2.7\Release\src\cache.o 
> build\temp.win32-2.7\Release\src\mic roprotocols.o 
> build\temp.win32-2.7\Release\src\prepare_protocol.o 
> build\temp.win32-2.7\Release\src\statement.o 
> build\temp.win32-2.7\Release\src\util.o 
> build\temp.win32-2.7\Release\src\row.o 
> build\temp.win32-2.7\Release\amalgamation\sqlite3 .o 
> build\temp.win32-2.7\Release\src\_sqlite.def -Lc:\openssl\lib 
> -Lc:\openssin -Lc:\zlib\lib -Lc:\zliin -LC:\Python27\libs 
> -LC:\Python27\PCbuild -LC:\Python27\PC\VS9.0 -lpython27 -lmsvcr90
> -o build\lib.win32-2.7\pysqlcipher\_sqlite.pyd lib eay32.lib 
> /LIBPATH:C:\lib gcc: error: libeay32.lib: No such file or directory
> gcc: error: /LIBPATH:C:\lib: No such file or directory error:
> command 'C:\\mingw\\bin\\gcc.exe' failed with exit status 1 die:
> setup.py for pysqlcipher failed Makefile:12: recipe for target
> 'pyinstaller' failed make[1]: *** [pyinstaller] Error 1 |
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/leapcode/bitmask_client/pull/965#issuecomment-1848
11072>.
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iQIcBAEBCAAGBQJWxaTzAAoJENwE0+YLGC5yX10P/21RNz6Rt3Xdhx3kiU9y2ftW
uPOtMYjO9JcHfazvLZduVOzp5TKmnwcULDKEFVBn7G5UN5J2jKyBvsHebZLbxcEX
cibKuFdsk+8XR7d9X5ZE0zVccybIBHsn+56/qG4EcbuGqaZJBpaeAwSuN1agfgyA
2BTDrgfWF3ZfRXnqDvX9Kvzsk8bEFskPbZdkpcqQn3p/N5fIKP5AJkrjmRfeTywV
itYvrwORH3dX7U4BzxsdEZyMBSkNRKbZejCbexyBAcXPEaiYDbf+U69k45Q6qyDq
u2YQgASmXVPtYVlVQJBof57k/JRjDdtgS1cDpsWGCbYiZc7AVFIKrNrNAMn92Hvw
CM6cZ+cHdd2fIn6aQ2SaYddmKFzq1my3WhMiCY/M6Ec8nLYAE4UUoXoSMJKfGjWG
sJs/kSvUpKe7883H3WNAhYTKw47ayti+NlQyrYU7c/VDhvg+ARst8rRZ2v6bYj2z
1FTQZpxQHZRWeKowyzOW5Kjns7uZqzqd4jU+gdupxJHOxMutUq7cGW92w9PMRbo5
ez0EjrpLmuoCo+1hyS651aaAwIub3vtvRQ7zyhIgD8zGyQf9zB2dOEX/6Kdk30xb
V+kiocBWXyYOYPCu0/sPbQzA7qPhr9n8R4DF/X+vPmaA6C8L8eKi5W1GG5JMfzSG
c1fb9xAljVC3x686R5jf
=wHuM
-----END PGP SIGNATURE-----

**Comment by meskio** [comment]
Great, now pysqlcipher builds and I got furder, but still hit an error:
https://share.riseup.net/#pyFY9bFagIcdinsTOJ7lmw

**Comment by PaixuAabuizia** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

https://github.com/giampaolo/psutil/issues/773
working on a solution...

On 02/18/2016 12:34 PM, Ruben Pollan wrote:
> Great, now pysqlcipher builds and I got furder, but still hit an
> error: https://share.riseup.net/#pyFY9bFagIcdinsTOJ7lmw
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/leapcode/bitmask_client/pull/965#issuecomment-1856
76086>.
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iQIcBAEBCAAGBQJWxazAAAoJENwE0+YLGC5yvuEP/25tMzaM+TGgLlQXua6dTzDp
L2yURTbV/zOiwbvbHNYbcTCsVqp0ozDiwBc6nR+m9mv0e8lMK34ZMnwSV1MQpQDE
lUvAQtogdZMmH0dJ21g2FEr/rS4bxywpxrbl9vVs/Dcc6PeWP53K+IK3tDAIo7zK
brc6UDkLnHNPHzRwYd6KNbITL4rIOWhIl9pGpZ9hmbWSrDZi8NEkHqAVzyzo55Z8
0Ktj21eyL8kerZtmFuB7zzEc38sSEzAGwN/pFYeYn83aEp91cBe8F8I9vaUyejL2
SgIKiym4oQ3CHxlcj7VY7NA9hugx2Z6Ee6diX3yfufYWlCclMPQ3O/5SANVn1xVq
crfCwMj7VywpdPJa0aIsNPOEJOFXhY3X7FNygJO8+VkAuaYTlrUV3FDmWhF2gfKX
zr8rKOj64ePExUTzwCrx6RiP0waVdO/p3PBhjxe+hXqtRS/tgYVW99ACiVJKdxaf
bgVcxszMaqta/7/xesMPL8aJKisVjloHSaRrKEBNdy+M196S95o5dtA2jATo7aeB
iYZsBEO+pXjxfcV3bcNl3nKv/1VUfGVzF1QGhaCL846nFT4aI6OgCvmT6V0JR+1V
F6pF5wwk4dyf3skG9cD05huzPtL4DbFeAzrExZLhWzuJ8TX4w+z9DRk/EkHNLX+6
Jysd1WzswEqsgefwfI5m
=Q4z9
-----END PGP SIGNATURE-----

File: pkg/pyinst/bitmask.spec (hunk 1 above)
Comment (PaixuAabuizia) [inline pkg/pyinst/bitmask.spec]: -----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

agreed, but pyinstaller fails with the following message if the
condition on sys.platform is in the spec file:

```
Traceback (most recent call last):
  File "C:\Python27\Scripts\pyinstaller-script.py", line 9, in <module>
    load_entry_point('PyInstaller==3.1.1', 'console_scripts',
'pyinstaller')()
  File "C:\python27\lib\site-packages\PyInstaller\__main__.py", line
90, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "C:\python27\lib\site-packages\PyInstaller\__main__.py", line
46, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs
)
  File
"C:\python27\lib\site-packages\PyInstaller\building\build_main.py",
line 755, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'),
kw.get('clean_build'))
  File
"C:\python27\lib\site-packages\PyInstaller\building\build_main.py",
line 701, in build
    exec(text, spec_namespace)
  File "<string>", line 47, in <module>
NameError: name 'sys' is not defined
```

On 02/12/2016 03:42 PM, Kali Kaneko wrote:
> In pkg/pyinst/bitmask.spec 
> <https://github.com/leapcode/bitmask_client/pull/965#discussion_r52748
544>:
>
> 
>> coll = COLLECT(exe, -               a.binaries, -
>> a.zipfiles, -               a.datas, -
>> strip=False, -               upx=True, -
>> name='bitmask') -if sys.platform.startswith("darwin"):
> 
> this shouldn't be removed, if we're going to use the same file for
> both win and osx builds
> 
> — Reply to this email directly or view it on GitHub 
> <https://github.com/leapcode/bitmask_client/pull/965/files#r52748544>.
>
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v2

iQIcBAEBCAAGBQJWxbKgAAoJENwE0+YLGC5y8AEP/RUubP7ufuLDBkvXZO1jdHpE
d15m8QJULeGQPU+SPRuCFrXqCtTmudh8sLOdBe34JIJPDG19e/2PHW5psUpCktxh
AJuTA3F1Bm7aWpCESYKbniSmqjcgQjzF/XdaLaoXmRTJHyfWMCPxyybEn305OOZS
9YjC10wjfuP+B4lpXrQ6tXRMs2mAG8hDcmhyLNVzUlC7s+/KQPIpAe70u9B9dnvl
8eNNlYkWCbP/VlfUU82Y7YwvExwh7f3gs8NhIdLs+f5eRCraVL/JxyB7cQtNUbtp
uBL1zLtxUIBVa4p1lR9J2N+T6YWHI49JoTr0SGOMSpbOM8i+P11obmd2ey5GwPJF
mfjkT35bc0oUA61i8Hj9mFTBB0BtoDLqfavGgPb//fidjdYVKGHfQKFztWlQ/B8O
VAtK7vlglvDAz8On26ffW9LoqdsN49e66mOPJWhIwT9xBORV2/SB/fszvVE9ebDC
UUoCaWiX1FN2udqIA7sW2R3ADVrvCAMRKTGN/RtN7V1ibe3+/oZqNoW18X19zlnZ
mJh+SwXtl7GG6fNfzoKuiPmkz/89gQFZEmUwKXHIW6b99x9xnk2RmjpK1yvq9zfO
9AdGrxCuI9Cv9Yxj7Aw2k4pN7hgdkPPS2P+r4i/wdyoJfn1XwxxLqvsRZB/Yz9dP
XBXhRB5n6z1b0ObEQV7J
=ak1S
-----END PGP SIGNATURE-----

**Comment by PaixuAabuizia** [comment]
updated the pyinstaller/Dockerfile to use psutils-3.4.2, this possibly breaks the docker-wine-installer approach.

**Comment by PaixuAabuizia** [comment]
```docker-compose build pyinstaller``` before the next ```docker-compose run --rm pyinstaller```

**Comment by meskio** [comment]
Now it worked, I have an installer \o/

The installer does install, but bitmask don't seem to work in my windows 7. If I launch it from PowerShell it returns immediately without any output. In bash (Mingw64) it just hangs again without output. 

The strace en Mingw64 (I don't know how to do something like that in PowerShell) output is: https://share.riseup.net/#dWmvigafXy9pTg5Subn9zg
It look like it got hang in the middle of loading win32pip.pyd.

Does it do anything in your set up?

**Comment by PaixuAabuizia** [comment]
what it does for me is:

```
Display Message Box "Bitmask returned -1"
```

see pkg/windows/TODO
the committed configuration is 'as windows software' ie no stdout /
immediate return (winmain).
to debug the installation:
edit pkg/pyinst/bitmask.spec
```
debug=True
console=True
```
then
```
linux> docker-compose run --rm pyinstaller
linux> cp -r ../../build/executables/pyinstaller $windows/c/debug
windows> cd c:\debug\ && bitmask -d
```

i did only get it to run with patches on
src/leap/bitmask/app.py (sleep 15s after backend_run - otherwise
backend is never up when frontend comes up -> shutdown)
change defaults in pkg/windows/pyinstaller-build.sh
(with_mail=false;with_eip=true)
src/leap/bitmask/platform_init/initializers.py:
to ```return True``` in WindowsInitializer

then start bitmask once
- fiddle with ~/AppData/leap/zmq/server.key (after you will see a
file-not-found exception)
start again (--danger?)
- see that zmq rejects the fiddled server key, but creates a new one
start again (--danger?)
- see the welcome wizard
if you loop on step 2, start over with rm -r ~/AppData/leap

with_mail=true it always crashed in the main_window init code, i did
not step into that, but something about accessing a member that does
not exist.

glad to read that the installer build code actually built something
installable on another machine!

**Comment by meskio** [comment]
The 'no output' problem might be related to a problem I also hit in my tests in windows, logbook (the logging library we are using) don't seems to work if we do 'perform_rollover'. See my crappy hacks to make bitmask working on windows:
https://github.com/meskio/bitmask_client/commit/0cbe5cd3011585b8f0b5d8d4251d462e1e650b9e

I need to get some of this properly commited in the main repo.

**Comment by meskio** [comment]
Ahh, I see now, that I didn't read your comment before replying, I'll try again activating the debug as you indicate. Thanks

File: pkg/pyinst/bitmask.spec (hunk 1 above)
Comment (kalikaneko) [inline pkg/pyinst/bitmask.spec]: @PaixuAabuizia  can you rebase against develop? 
this is fixed now, there's a "import sys" in there now

**Comment by kalikaneko** [comment]
with the caveat of the pyinstaller spec, that simply should be rebased, i'm ok with merging this, great work!

**Comment by kalikaneko** [comment]
with the caveat of the pyinstaller spec, that simply should be rebased, i'm ok with merging this, great work!

File: pkg/windows/pyinstaller-build.sh [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  # find the information and put it into the egg
+  versioned_build_path=/var/tmp/${version_prefix}-${git_version}
+  mkdir -p ${versioned_build_path}
+  cp -r ${temporary_build_path}/* ${versioned_build_path}
+  # apply patches to the source that are required for working code
+  # should not be required in the future as it introduces possible
+  # hacks that are hard to debug
+  applyPatches ${versioned_build_path}
+  pushd ${versioned_build_path} > /dev/null
+  wine python setup.py update_files || die 'setup.py update_files failed'
```
Comment (kalikaneko) [inline pkg/windows/pyinstaller-build.sh]: what's this command? it fails for me, setup.py doesn't contain it.

**Comment by kalikaneko** [comment]
I rebased and merged this branch before the 0.9.2 tag

**Pull request closed without merging**